File: spring-web-java/src/main/java/spring/web/java/domain/member/serivce/MemberService.java
package spring.web.java.domain.member.serivce;

import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;
import spring.web.java.domain.Address;
import spring.web.java.domain.member.Member;
import spring.web.java.domain.member.dto.MemberResponse;
import spring.web.java.domain.member.dto.MemberSaveRequest;
import spring.web.java.domain.member.dto.MemberUpdateRequest;
import spring.web.java.domain.member.repository.MemberRepository;
import spring.web.java.global.common.ResponseMessage;
import spring.web.java.global.common.SecurityService;
import spring.web.java.global.exception.DomainException;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class MemberService {

	private final MemberRepository memberRepository;
	private final PasswordEncoder passwordEncoder;
	private final SecurityService securityService;

	@Transactional
	public MemberResponse saveMember(MemberSaveRequest memberSaveRequest) {
		memberRepository.findByAccount(memberSaveRequest.account())
			.ifPresent(member -> {
				throw new DomainException(ResponseMessage.DUPLICATE_DATA, HttpStatus.CONFLICT);
			});

		Member member = memberRepository.save(
			Member.create(
				memberSaveRequest.account(),
				passwordEncoder.encode(memberSaveRequest.password()),
				memberSaveRequest.name(),
				memberSaveRequest.role(),
				Address.create(
					memberSaveRequest.city(),
					memberSaveRequest.street(),
					memberSaveRequest.zipcode()
				)
			)
		);

		return new MemberResponse(member);
	}

	public MemberResponse findMember() {
		return memberRepository.findById(securityService.getMemberId())
			.map(MemberResponse::new)
			.orElseThrow(() -> new DomainException(ResponseMessage.DATA_NOT_FOUND, HttpStatus.NOT_FOUND));
	}

	@Transactional
	public MemberResponse updateMember(MemberUpdateRequest memberUpdateRequest) {
		Member member = memberRepository.findById(securityService.getMemberId())
			.orElseThrow(() -> new DomainException(ResponseMessage.DATA_NOT_FOUND, HttpStatus.NOT_FOUND));

		member.update(
			memberUpdateRequest.name(),
			memberUpdateRequest.role(),
			Address.create(
				memberUpdateRequest.city(),
				memberUpdateRequest.street(),
				memberUpdateRequest.zipcode()
			)
		);

		return new MemberResponse(member);
	}

	@Transactional
	public void deleteMember() {
		memberRepository.deleteById(securityService.getMemberId());
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/member/service/MemberService.kt
package spring.web.kotlin.domain.member.service

import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import spring.web.kotlin.domain.Address
import spring.web.kotlin.domain.member.Member
import spring.web.kotlin.domain.member.dto.MemberResponse
import spring.web.kotlin.domain.member.dto.MemberSaveRequest
import spring.web.kotlin.domain.member.dto.MemberUpdateRequest
import spring.web.kotlin.domain.member.repository.MemberRepository
import spring.web.kotlin.global.common.ResponseMessage
import spring.web.kotlin.global.common.SecurityUtils
import spring.web.kotlin.global.exception.BaseException

@Service
@Transactional(readOnly = true)
class MemberService(
    private val memberRepository: MemberRepository,
    private val passwordEncoder: PasswordEncoder
) {
    @Transactional
    fun saveMember(memberSaveRequest: MemberSaveRequest): MemberResponse {
        memberRepository.findByAccount(memberSaveRequest.account)?.let {
            throw BaseException(ResponseMessage.DUPLICATE_DATA, HttpStatus.CONFLICT)
        }

        val member = memberRepository.save(
            Member.create(
                account = memberSaveRequest.account,
                password = passwordEncoder.encode(memberSaveRequest.password),
                name = memberSaveRequest.name,
                role = memberSaveRequest.role,
                address = Address.create(
                    city = memberSaveRequest.city,
                    street = memberSaveRequest.street,
                    zipcode = memberSaveRequest.zipcode,
                ),
            ),
        )

        return MemberResponse(member)
    }

    fun findMember(): MemberResponse {
        val id = SecurityUtils.getMemberId()
            ?: throw BaseException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED)

        val member = memberRepository.findByIdOrNull(id)
            ?: throw BaseException(ResponseMessage.DATA_NOT_FOUND, HttpStatus.NOT_FOUND)

        return MemberResponse(member)
    }

    @Transactional
    fun patchMember(memberUpdateRequest: MemberUpdateRequest): MemberResponse {
        val id = SecurityUtils.getMemberId()
            ?: throw BaseException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED)

        val member = memberRepository.findByIdOrNull(id)
            ?: throw BaseException(ResponseMessage.DATA_NOT_FOUND, HttpStatus.NOT_FOUND)

        member.update(
            name = memberUpdateRequest.name,
            role = memberUpdateRequest.role,
            address =
            Address.create(
                city = memberUpdateRequest.city,
                street = memberUpdateRequest.street,
                zipcode = memberUpdateRequest.zipcode,
            ),
        )

        return MemberResponse(member)
    }

    @Transactional
    fun deleteMember() {
        val id = SecurityUtils.getMemberId()
            ?: throw BaseException(ResponseMessage.AUTHENTICATION_FAILED, HttpStatus.UNAUTHORIZED)

        memberRepository.deleteById(id)
    }
}
